src/expr/src/scalar.rs

```diff
@@ -1231,15 +1231,29 @@ impl MirScalarExpr {
 
                             // Defer errors until evaluation instead of eagerly returning them here
                             // to match the error behavior of the dynamic function (part of database-issues#4972).
-                            let regex = match func::build_regex(pattern, &flags) {
-                                Ok(regex) => Ok((regex, limit)),
-                                Err(err) => Err(err),
+                            *e = match func::build_regex(pattern, &flags) {
+                                Ok(regex) => {
+                                    let mut exprs = mem::take(exprs);
+                                    let replacement = exprs.swap_remove(2);
+                                    let source = exprs.swap_remove(0);
+                                    source.call_binary(
+                                        replacement,
+                                        BinaryFunc::RegexpReplace { regex, limit },
+                                    )
+                                }
+                                Err(err) => {
+                                    let mut exprs = mem::take(exprs);
+                                    let source = exprs.swap_remove(0);
+                                    let scalar_type = e.typ(column_types).scalar_type;
+                                    // We need to return `NULL` on `NULL` input, and error otherwise.
+                                    source
+                                        .call_is_null()
+                                        .if_then_else(
+                                            MirScalarExpr::literal_null(scalar_type.clone()),
+                                            MirScalarExpr::literal(Err(err), scalar_type),
+                                        )
+                                }
```

Comment:
@ggevay I changed the error path to return `null` on `null` input, and error otherwise. I think this should match the current behavior!